SUBMARINE-469. Update Submarine Architecture Doc

wangdatan opened the pull request:
### What is this PR for?
We have many discussions recently and made some proposals for the Submarine roadmap and corresponding roadmaps. We need to update our docs.


### What type of PR is it?
Documentation

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-469

### How should this be tested?
* First time? Setup Travis CI as described on https://submarine.apache.org/contribution/contributions.html#continuous-integration
* Strongly recommended: add automated unit tests for any new or changed behavior
* Outline any manual steps to test the PR here.

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? Yes/No
* Is there breaking changes for older versions? Yes/No
* Does this needs documentation? Yes/No

cchung100m commented:
Hi @wangdatan 

Thanks for contributing to documents.

Please help to re-trigger the CI process for solving the failed job.

many thanks,

pingsutw commented:
+1, Thanks @wangdatan. The new design doc is very clear to me.

wangdatan commented:
Thanks for the reviews, I will spend a bit more time to finish a few more areas before committing the patch. Please feel free to add comments/questions when you have time.

liuxunorg commented on docs/design/environments-implementation.md:
````diff
@@ -0,0 +1,193 @@
+
+```
+    name: "my_submarine_env",
+    vm-image: "...",
````
I do n’t quite understand the meaning of `vm-image`, Does `vm-image` refer to the ec2 machine type on aws for example?
https://amazonaws-china.com/cn/ec2/instance-types/
But the environment of ec2 server cannot be provided by docker image.

wangdatan replied:
It is different from the instance-type.  
I just added a note to the same doc: 

> Docker and/or VM-image (such as, VirtualBox/VMWare images, Amazon Machine Images - > AMI, Or custom image of Azure VM) defines the base layer of the environment. Please note > that VM-image is different from VM instance type,

liuxunorg commented on docs/design/environments-implementation.md:
```diff
@@ -0,0 +1,193 @@
+<Make sure conda (with our preferred version) is installed>
+<Make sure Jupyter (with our preferred version) is installed>
+
+ENTRYPOINT ["/submarine-bootstrap.sh"]
```
ENTRYPOINT ["/submarine-bootstrap.sh"] # This line is a sample code, the actual situation may change a little

wangdatan replied:
Updated doc and made it more clear.

liuxunorg commented on docs/design/experiment-implementation.md:
```diff
@@ -0,0 +1,248 @@
+| --------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| Run mode                          | Offline                                                      | Interactive                                                  |
+| Output Artifacts (a.k.a model)    | Persisted in a shared storage (like S3/NFS)                  | Local in the notebook session container, could be emphameral |
+| Run history (meta, logs, metrics) | Meta/logs/metrics can be traced from experiment UI (or corresponding API) | No run history can be traced from Submarine UI/API           |
```
After `No run history can be traced from Submarine UI/API`, add `Can view the current running log.`

wangdatan replied:
Updated

liuxunorg commented on docs/design/storage-implementation.md:
```diff
@@ -0,0 +1,157 @@
+
+- Notebook 
+- Experiment
+- Endpoint
```
`Endpoint`, I think it's clearer to just call `Serving`.

liuxunorg commented on docs/design/storage-implementation.md:
```diff
@@ -0,0 +1,157 @@
+      +----------+            +---------+    +------------+     +----------------+
+      |Trackings |                        <-+|Experiment  |<--+>|Model Artifacts |
+      +----------+     +-----------------+   +------------+     +----------------+
+      +----------+<---+|ML-related Metric|<--+Endpoint    |
```
`Endpoint`, I think it's clearer to just call `Serving`.

liuxunorg commented on docs/design/storage-implementation.md:
```diff
@@ -0,0 +1,157 @@
+
+Making code to be part of Docker image makes hard to edit code (if you want to update a value in your Python file, you will have to recreate the Docker image, push it and rerun it).
+
+**2) Code is part of Git repo:** 
```
Because this is our recommended way, it is better to put it on the first line.
**1) Code is part of Git repo:** (recommended)

wangdatan replied:
Make sense, updated.

wangdatan commented on docs/design/architecture-and-requirements.md:
```diff
@@ -142,164 +137,211 @@ Following items are charters of Submarine project:
+- Users can choose to save, tag, version a produced model.
+- Once The Model is saved, Users can do the online serving (endpoint) or offline scoring of the model.
+
+#### Model serving (Endpoint)
```
Make sense, updating.

wangdatan commented on docs/design/architecture-and-requirements.md:
```diff
@@ -142,164 +137,211 @@ Following items are charters of Submarine project:
+
+     +--------------------Submarine Server-----------------------------+
+     | +---------+ +---------+ +----------+ +----------+ +------------+|
+     | |Data set | |Notebooks| |Experiment| |Models    | |endpoints   ||
```
Updated

wangdatan commented:
Changes added since Friday: 
- Updated the main architecture doc to include fewer details of implementation.
- Included a lot more details of module designs for experiment inside experiment-implementation.md. Also included thoughts about how to support pre-defined templates in implementation. @pingsutw / @liuxunorg / @tangzhankun / @jiwq  can you help to check them? 
- Removed implementation for experiment related module discussion from submarine-server/architecture to experiment-implementation.
- it looks like submarine-launcher.md and SubmarineClusterServer.md are not complete or obsolete. I moved them to design/wip-designs. @liuxunorg please correct me if I was wrong.

liuxunorg commented:
> it looks like submarine-launcher.md and SubmarineClusterServer.md are not complete or obsolete. I moved them to design/wip-designs. @liuxunorg please correct me if I was wrong.

@wangdatan  Yes, you are right.

liuxunorg commented:
Will merge if no more comments.
